src/securityPolicy.c

```diff
@@ -565,86 +656,124 @@ security_policy_t policyForSignTxStakePoolRegistrationConfirm(
 }
 
 // For each withdrawal
-security_policy_t policyForSignTxWithdrawal()
+security_policy_t policyForSignTxWithdrawal(
+        sign_tx_signingmode_t txSigningMode,
+        const stake_credential_t* stakeCredential
+)
 {
-	// No need to check withdrawals
-	SHOW();
-}
+	switch (txSigningMode) {
+	case SIGN_TX_SIGNINGMODE_ORDINARY_TX:
+		DENY_UNLESS(stakeCredential->type == STAKE_CREDENTIAL_KEY_PATH);
+		DENY_UNLESS(bip44_isOrdinaryStakingKeyPath(&stakeCredential->keyPath));
+		ALLOW();
+		break;
 
+	case SIGN_TX_SIGNINGMODE_SCRIPT_TX:
+		DENY_UNLESS(stakeCredential->type == STAKE_CREDENTIAL_SCRIPT_HASH);
+		ALLOW();
+		break;
 
-// For each transaction witness
-// Note: witnesses reveal public key of an address
-// and Ledger *does not* check whether they correspond to
-// previously declared inputs and certificates
-security_policy_t policyForSignTxWitness(
-        sign_tx_usecase_t signTxUsecase,
-        const bip44_path_t* pathSpec
-)
+	default:
+		ASSERT(false);
+	}
+
+	DENY(); // should not be reached
+}
+
+static inline security_policy_t _ordinaryWitnessPolicy(const bip44_path_t* path, bool mintPresent)
 {
-	switch (signTxUsecase) {
+	switch (bip44_classifyPath(path)) {
+	case PATH_ORDINARY_SPENDING_KEY:
+	case PATH_ORDINARY_STAKING_KEY:
+	case PATH_POOL_COLD_KEY:
```

Comment:
should we really allow pool cold keys in the "ordinary" case? Isn't that reserved only for the case when signing a pool registration as an owner?

Reply:
They are also used for pool retirement certificates that are only included in "ordinary" transactions. But there is no reason to allow it for script tx, so I've removed it from there.